src/experiments/logins/api.js

```diff
@@ -84,32 +84,19 @@ this.logins = class extends ExtensionAPI {
               throw new ExtensionError(ex);
             }
           },
-          update(newLogin) {
-            if (!("guid" in newLogin)) {
+          update(updatedLogin) {
+            if (!("guid" in updatedLogin)) {
               throw new ExtensionError(`Update failed: ID is required`);
             }
-            // The addons framework code always supplies null values for missing keys,
-            // so filter those out.
-            // TODO: Could this cause subtle bugs? Would anyone ever want to update a field to be null? If so, maybe default to 'NULL" (the string), or undefined.
-            for (let key in newLogin) {
-              if (newLogin[key] === null) {
-                delete newLogin[key];
-              }
-            }
-            const oldLogin = getLogin(newLogin.guid);
+            const oldLogin = getLogin(updatedLogin.guid);
             if (!oldLogin) {
-              throw new ExtensionError(`Update failed: Login not found with ID ${newLogin.guid}`);
+              throw new ExtensionError(`Update failed: Login not found with ID ${updatedLogin.guid}`);
             }
             try {
-              const updatedLogin = {};
-              Object.assign(updatedLogin, oldLogin, newLogin);
-              if ("timesUsedIncrement" in newLogin) {
-                delete updatedLogin.timesUsed;
-              }
-              const updatedLoginBag = LoginHelper.newPropertyBag(updatedLogin);
+              const updatedLoginInfo = LoginHelper.vanillaObjectToLogin(updatedLogin);
```

Comment:
I scratched my head at the description of this issue for a little bit. But, I think the crux of the issue is where [modifyLogin()](https://developer.mozilla.org/en-US/docs/Mozilla/Tech/XPCOM/Reference/Interface/nsILoginManager#modifyLogin()) can accept either an nsIPropertyBag or an nsILoginMetaInfo instance.

Switched to the latter, here, on assumption of a full login object modified by the calling code as necessary, rather than a sparse set of updates. So I *think* that's me understanding the issue correctly.

Reply:
I think it'd be better to stick with the PropertyBag as the second argument to `modifyLogin`.

Sorry this wasn't clearer from the bug description. The LoginInfo fields can be updated in this way, but not the [LoginMetaInfo](https://developer.mozilla.org/en-US/docs/Mozilla/Tech/XPCOM/Reference/Interface/nsILoginMetaInfo) fields (timestamps and usage count). As I understand it, the only reason there's a separate interface for the MetaInfo fields is that they were needed after login sync was already done, so each login object actually implements both interfaces: the syncable part, and the non-syncable part. Even with the historical background in mind, it's a confusing and, uh, suboptimal situation.

We could just rely on the `touch()` method to alter the MetaInfo fields, but that makes `update()` an annoyingly misleading function name, and causes the grody dual-interface situation to complicate our lives.

Reply:
Switched back to a property bag. But, it looks like we might have to rely on `touch()` for using the `timesUsedIncrement` property.

Reply:
relying on `touch()` seems reasonable to me.